### PR TITLE
[Flax] ViT training example

### PR DESCRIPTION
**Pull request by patil-suraj**
# What does this PR do?

This PR adds image classification script for fine-tuning flax vit.

For faster processing and loading , torch dataloader is used, since this can become a bottleneck on TPU.

### CODE REVIEW

File: examples/flax/vision/run_image_classification.py [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+has_tensorboard = is_tensorboard_available()
+if has_tensorboard:
+    try:
+        from flax.metrics.tensorboard import SummaryWriter
+    except ImportError as ie:
+        has_tensorboard = False
+        print(f"Unable to display metrics through TensorBoard because some package are not installed: {ie}")
+
+else:
+    print(
```
Comment (patrickvonplaten) [inline examples/flax/vision/run_image_classification.py:76]: Maybe put inside the `main()` function and user logger instead

File: examples/flax/vision/run_image_classification.py [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+    else:
+        datasets.utils.logging.set_verbosity_error()
+        transformers.utils.logging.set_verbosity_error()
+
+    # Set the verbosity to info of the Transformers logger (on main process only):
+    logger.info(f"Training/evaluation parameters {training_args}")
+
+    # Load pretrained model and tokenizer
+
+    # Distributed training:
```
Comment (patrickvonplaten) [inline examples/flax/vision/run_image_classification.py:255]: can be deleted

File: examples/flax/vision/run_image_classification.py [hunk 3]
```diff
@@ -0,0 +1,486 @@
...
+                normalize,
+            ]
+        ),
+    )
+
+    eval_dataset = torchvision.datasets.ImageFolder(
+        data_args.valid_dir,
+        transforms.Compose(
+            [
+                transforms.Resize(256),
```
Comment (patrickvonplaten) [inline examples/flax/vision/run_image_classification.py:295]: Is it normal that the transforms compose is different to the one in the training set here?

File: examples/flax/vision/run_image_classification.py [hunk 4]
```diff
@@ -0,0 +1,486 @@
...
+        eps=training_args.adam_epsilon,
+        weight_decay=training_args.weight_decay,
+        mask=decay_mask_fn,
+    )
+
+    # Setup train state
+    state = TrainState.create(apply_fn=model.__call__, params=model.params, tx=adamw, dropout_rng=dropout_rng)
+
+    def loss_fn(logits, labels):
+        shift_logits = logits[..., :-1, :]
```
Comment (patrickvonplaten) [inline examples/flax/vision/run_image_classification.py:379]: this needs to be corrected as discussed

File: examples/flax/vision/run_image_classification.py [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+        training_args.learning_rate,
+    )
+
+    # We use Optax's "masking" functionality to not apply weight decay
+    # to bias and LayerNorm scale parameters. decay_mask_fn returns a
+    # mask boolean with the same structure as the parameters.
+    # The mask is True for parameters that should be decayed.
+    def decay_mask_fn(params):
+        flat_params = traverse_util.flatten_dict(params)
+        flat_mask = {path: (path[-1] != "bias" and path[-2:] != ("LayerNorm", "scale")) for path in flat_params}
```
Comment (patrickvonplaten) [inline examples/flax/vision/run_image_classification.py:362]: this needs to be corrected as well

File: examples/flax/vision/run_image_classification.py [hunk 6]
```diff
@@ -0,0 +1,486 @@
...
+
+    # Store some constant
+    num_epochs = int(training_args.num_train_epochs)
+    train_batch_size = int(training_args.per_device_train_batch_size) * jax.device_count()
+    eval_batch_size = int(training_args.per_device_eval_batch_size) * jax.device_count()
+    steps_per_epoch = len(train_dataset) // train_batch_size
+    total_train_steps = steps_per_epoch * num_epochs
+
+    def collate_fn(examples):
+        pixel_values = torch.stack([example[0] for example in examples])
```
Comment (patrickvonplaten) [inline examples/flax/vision/run_image_classification.py:311]: Just to be sure `exmaple[1]` is really the labels no? not the attention mask or something?

File: examples/flax/vision/run_image_classification.py [hunk 7]
```diff
@@ -0,0 +1,486 @@
...
+        data_args.valid_dir,
+        transforms.Compose(
+            [
+                transforms.Resize(256),
+                transforms.CenterCrop(224),
+                transforms.ToTensor(),
+                normalize,
+            ]
+        ),
+    )
```
Comment (patrickvonplaten) [inline examples/flax/vision/run_image_classification.py:301]: Can we verify somehow that the data is correctly processed here? Some simple test?

**Comment by patrickvonplaten** [comment]
Can you run `make style` ?

File: examples/flax/vision/README.md [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    --train_dir="imagenette2/train" \
+    --validation_dir="imagenette2/val" \
+    --num_train_epochs 5 \
+    --learning_rate 1e-3 \
+    --per_device_train_batch_size 128 --per_device_eval_batch_size 128 \
+    --overwrite_output_dir \
+    --preprocessing_num_workers 32
+```
+
+This should finish in ~7mins with 99% validation accuracy.
```
Comment (patrickvonplaten) [inline examples/flax/vision/README.md:100]: Can we add `--push_to_hub` here?

**Review by patrickvonplaten — APPROVED** [review]
